types/logger/logger.go

```diff
@@ -42,3 +46,61 @@ func (w funcWriter) Write(p []byte) (int, error) {
 
 // Discard is a Logf that throws away the logs given to it.
 func Discard(string, ...interface{}) {}
+
+type limitData struct {
+	lim          *rate.Limiter
+	lastAccessed time.Time
+	msgBlocked   bool
+}
+
+// RateLimitedFn implements rate limiting by fstring on a given Logf.
+// Messages are allowed through at a maximum of r messages/second, in
+// bursts of up to b messages at a time.
+func RateLimitedFn(logf Logf, f float64, b int) Logf {
+	r := rate.Limit(f)
+	msgList := make(map[string]limitData)
+	lastPurge := time.Now()
+
+	rlLogf := func(s string, args ...interface{}) {
+		if rl, ok := msgList[s]; ok {
+
+			// Fields of structs contained in maps can't be modified; this is
+			// the workaround. See issue https://github.com/golang/go/issues/3117
+			temp := msgList[s]
+			temp.lastAccessed = time.Now()
+			msgList[s] = temp
+
+			if rl.lim.Allow() {
+				rl.msgBlocked = false
+				logf(s, args)
+			} else {
+				if !rl.msgBlocked {
+					temp = msgList[s]
+					temp.msgBlocked = true
+					msgList[s] = temp
+					logf("Repeated messages were suppressed by rate limiting. Original message: " +
+						fmt.Sprintf(s, args))
+				}
+			}
+		} else {
+			msgList[s] = limitData{rate.NewLimiter(r, b), time.Now(), false}
+			msgList[s].lim.Allow()
+			logf(s, args)
+		}
+
+		// Purge msgList of outdated keys to reduce overhead. Must be done by copying
+		// over to a new map, since deleting in maps is done through a zombie flag
```

Comment:
I don't follow this. Why can't we just delete from the map?

Reply:
I was reading up on [stackoverflow](https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-map-within-a-range-loop) about this, apparently in go when you call `delete` on a map entry it doesn't actually free up the memory for the garbage collector, it just zombie flags it as "no longer in the map."
To make it actually available for the garbage collector you have to delete the map altogether, which is why I'm just copying over the values we still want to keep track of into a new map and setting the original to `nil` to be more clear about what I was doing - though I can just comment this in and clean up the code.
I might change how the purging is done entirely with the LRU Avery suggested though so this might go poof altogether :)

Reply:
I understand what you're saying about delete not cleaning up the map, but it's not as serious as you think. the "zombie" flag basically means that slot in the map is recyclable, ie. it can be reused by some future map entry. The array containing the map might never shrink, but as long as you make sure there are never more than N entries in the map, the map will reach a certain maximum size and stay there. Luckily, the LRU can make exactly that guarantee.

Reply:
Oh if that's the case then I've removed the expensive copying parts, it's just deleting instead.